### PR TITLE
`Integrated code lifecycle`: Add auxiliary repository view

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

It is possible to view all repositories in the repository view, except auxiliary repositories. 

### Description

This PR adds the capability to view the auxiliary repository.

Disclaimer: as this PR makes changes on various levels of the code-editor, we could refactor it to adhere to the new client guidelines. 
We will refactor the code editor to use the new client guidelines in a follow-up, as the change is very drastic.

### Steps for Testing

Prerequisites:
- 1 Instructor
1. Create a programming exercise and add an auxiliary repository.
2. Clone the auxiliary repository and commit something.
3. In the exercise details page, use the code button to go to the auxiliary repositories repository view (the link icon on the right of the link)
image

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| programming-exercise.service.ts | 90.18% | ✅ ❌ |
| code-editor.model.ts | 100% | ✅ |
| code-editor-conflict-state.service.ts | 96.87% | ✅ ❌ |
| code-editor-domain.service.ts | 100% | ✅ |
| commit-history.component.ts | 100% | ✅ |
| repository-view.component.ts | 98.03% | ✅ ❌ |

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| ProgrammingExerciseParticipationService.java | 74% | ✅ ❌ |
| ProgrammingExerciseService.java | 87% | ✅ ❌ |
| ProgrammingExerciseParticipationResource.java | 88% | ✅ ❌ |
| ProgrammingExerciseResource.java | 79% | ✅ ❌ |
| AuxiliaryRepositoryResource.java | 97% | ✅ ❌ |

### Screenshots

image

## Release Notes

- **New Features**
  - Introduced a comprehensive set of RESTful endpoints for managing auxiliary repositories, allowing users to perform actions such as creating, deleting, and renaming files and folders.
  - Added functionality for pulling changes, committing changes, and checking repository status.
  - Enhanced commit history retrieval specifically for auxiliary repositories.
  - Implemented a new method to retrieve programming exercises along with their associated auxiliary repositories.
  - Added routing configurations for improved access to repository-related views.

- **Bug Fixes**
  - Improved error handling for repository operations, ensuring robust management of access violations and state conflicts.

- **Improvements**
  - Enhanced security measures for endpoint access, ensuring only authorized users can perform repository actions.
  - Refactored commit handling logic for better organization and clarity in the application.

- **Documentation**
  - Updated documentation to reflect new functionalities and endpoint usage for auxiliary repositories.

### CODE REVIEW

**Review by coolchock — APPROVED** [review]
Tested on TS3, works as expected. The Aux repo code is visible in the editor, and the commit history functions correctly.

For a follow-up: We might consider improving the design for listing Aux repos. Currently, if we have two repos with different name lengths, the buttons are not aligned.

File: src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseResource.java [hunk 1]
```diff
@@ -531,6 +531,26 @@ public ResponseEntity<ProgrammingExercise> getProgrammingExerciseWithTemplateAnd
...
+     *
+     * @param exerciseId            the id of the programmingExercise to retrieve
+     * @param auxiliaryRepositoryId the id of the auxiliary repository of the exercise
+     * @return the ResponseEntity with status 200 (OK) and the programming exercise with template and solution participation, or with status 404 (Not Found)
+     */
+    @GetMapping("programming-exercises/{exerciseId}/with-auxiliary-repository/{auxiliaryRepositoryId}")
+    @EnforceAtLeastTutorInExercise
+    public ResponseEntity<ProgrammingExercise> getProgrammingExerciseWithTemplateAndSolutionParticipation(@PathVariable long exerciseId, @PathVariable long auxiliaryRepositoryId) {
+
+        log.debug("REST request to get programming exercise with template and solution participation : {}", exerciseId);
```
Comment (pzdr7) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseResource.java:545]: Could you update the method name and log to reflect the endpoint? I feel like it should mention the auxiliary repository

File: src/main/webapp/app/detail-overview-list/detail-overview-list.component.html [hunk 2]
```diff
@@ -4,7 +4,7 @@
 @for (section of sections; track section) {
     <h3 class="section-headline" [id]="headlinesRecord[section.headline]">{{ section.headline | artemisTranslate }}</h3>
     <dl class="section-detail-list">
-        @for (detail of section.details; track detail) {
+        @for (detail of section.details; track $index) {
```
Comment (pzdr7) [inline src/main/webapp/app/detail-overview-list/detail-overview-list.component.html:7]: Was this change intentional for this PR?

File: src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts [hunk 3]
```diff
@@ -185,4 +185,10 @@ export class ProgrammingExerciseParticipationService implements IProgrammingExer
     retrieveCommitHistoryForTemplateSolutionOrTests(exerciseId: number, repositoryType: string): Observable<CommitInfo[]> {
         return this.http.get<CommitInfo[]>(`${this.resourceUrl}${exerciseId}/commit-history/${repositoryType}`);
     }
+
+    retrieveCommitHistoryForAuxiliaryRepository(exerciseId: number, repositoryType: string, auxiliaryRepositoryId: number): Observable<CommitInfo[]> {
+        const params: { [key: string]: number | string } = {};
```
Comment (pzdr7) [inline src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts:190]: Would this work here?

```suggestion
        const params: { [key: string]: number } = {};
```

File: src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts [hunk 4]
```diff
@@ -185,4 +185,10 @@ export class ProgrammingExerciseParticipationService implements IProgrammingExer
     retrieveCommitHistoryForTemplateSolutionOrTests(exerciseId: number, repositoryType: string): Observable<CommitInfo[]> {
         return this.http.get<CommitInfo[]>(`${this.resourceUrl}${exerciseId}/commit-history/${repositoryType}`);
     }
+
+    retrieveCommitHistoryForAuxiliaryRepository(exerciseId: number, repositoryType: string, auxiliaryRepositoryId: number): Observable<CommitInfo[]> {
```
Comment (pzdr7) [inline src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts:189]: Do we need the `repositoryType` here? It looks unused

File: src/main/java/de/tum/cit/aet/artemis/programming/web/repository/AuxiliaryRepositoryResource.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+import de.tum.cit.aet.artemis.programming.repository.ProgrammingExerciseRepository;
+import de.tum.cit.aet.artemis.programming.service.GitService;
+import de.tum.cit.aet.artemis.programming.service.RepositoryAccessService;
+import de.tum.cit.aet.artemis.programming.service.RepositoryService;
+import de.tum.cit.aet.artemis.programming.service.localvc.LocalVCRepositoryUri;
+import de.tum.cit.aet.artemis.programming.service.localvc.LocalVCServletService;
+import de.tum.cit.aet.artemis.programming.service.vcs.VersionControlService;
+
+/**
+ * Executes requested actions on the test repository of a programming exercise. Only available to TAs, Instructors and Admins.
```
Comment (pzdr7) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/repository/AuxiliaryRepositoryResource.java:56]:
```suggestion
 * Executes requested actions on the auxiliary repositories of a programming exercise. Only available to TAs, Instructors and Admins.
```

File: src/main/java/de/tum/cit/aet/artemis/programming/web/repository/AuxiliaryRepositoryResource.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+    @Override
+    @GetMapping(value = "auxiliary-repository/{auxiliaryRepositoryId}", produces = MediaType.APPLICATION_JSON_VALUE)
+    @EnforceAtLeastTutor
+    public ResponseEntity<RepositoryStatusDTO> getStatus(@PathVariable Long auxiliaryRepositoryId) throws GitAPIException {
+        return super.getStatus(auxiliaryRepositoryId);
+    }
+
+    /**
+     * Update a list of files in a test repository based on the submission's content.
```
Comment (pzdr7) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/repository/AuxiliaryRepositoryResource.java:184]:
```suggestion
     * Update a list of files in an auxiliary repository based on the submission's content.
```

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Please also add some tests for the new functions

File: src/main/webapp/app/detail-overview-list/detail-overview-list.component.html (hunk 2 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/detail-overview-list/detail-overview-list.component.html:7]: Yes, this is actually a bugfix. Interestingly, you only encounter the issue locally though. When you go to the execise-details page, this warning is constantly triggered:

image

File: src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts (hunk 4 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts:189]: Am I missing something? I think it gets used, in line 192, at the very end before the params
```
        return this.http.get<CommitInfo[]>(`${this.resourceUrl}${exerciseId}/commit-history/${repositoryType}`, { params: params });
```

Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts:189]: Sorry, I was looking at the wrong method. Would you say that `repositoryType` is required as a parameter here given the method name `retrieveCommitHistoryForAuxiliaryRepository`? I feel like it would only ever be `AUXILIARY` here

Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/manage/services/programming-exercise-participation.service.ts:195]: aah I see yea, it's indeed not needed here :+1:

**Review by sarpsahinalp — DISMISSED** [review]
Tested locally seems to be working as described!

File: src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseResource.java [hunk 7]
```diff
@@ -531,6 +531,26 @@ public ResponseEntity<ProgrammingExercise> getProgrammingExerciseWithTemplateAnd
...
+    /**
+     * GET /programming-exercises/:exerciseId/with-auxiliary-repository/:auxiliaryRepositoryId
+     *
+     * @param exerciseId            the id of the programmingExercise to retrieve
+     * @param auxiliaryRepositoryId the id of the auxiliary repository of the exercise
+     * @return the ResponseEntity with status 200 (OK) and the programming exercise with template and solution participation, or with status 404 (Not Found)
+     */
+    @GetMapping("programming-exercises/{exerciseId}/with-auxiliary-repository/{auxiliaryRepositoryId}")
+    @EnforceAtLeastTutorInExercise
+    public ResponseEntity<ProgrammingExercise> getProgrammingExerciseWithAuxiliaryRepository(@PathVariable long exerciseId, @PathVariable long auxiliaryRepositoryId) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseResource.java:543]: I'd assume that this endpoint only returns the exercise with the requested auxiliary repository but apparently all related aux repos are returned? In this case, why do you include the aux repo id at all?

File: src/main/webapp/app/localvc/repository-view/repository-view.component.ts [hunk 8]
```diff
@@ -190,4 +194,28 @@ export class RepositoryViewComponent implements OnInit, OnDestroy {
...
+                }),
+            )
+            .subscribe({
+                next: () => {
+                    this.loadingParticipation = false;
+                },
+                error: () => {
+                    this.participationCouldNotBeFetched = true;
+                    this.loadingParticipation = false;
+                },
```
Comment (JohannesStoehr) [inline src/main/webapp/app/localvc/repository-view/repository-view.component.ts:218]:
```suggestion
                complete: () => {
                    this.loadingParticipation = false;
                },
                error: () => {
                    this.participationCouldNotBeFetched = true;
                },
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Looks good overall, left two remarks:

File: src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseResource.java (hunk 7 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseResource.java]: Good point, that's indeed redundant here. I removed th erepository Id here.

**Review by farisd16 — APPROVED** [review]
Tested on TS3, works as expected

**Review by edkaya — DISMISSED** [review]
tested on ts3, works as expected

**Review by cremertim — DISMISSED** [review]

**Review by JohannesStoehr — DISMISSED** [review]
Code

File: src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+import de.tum.cit.aet.artemis.programming.dto.RepositoryStatusDTO;
+import de.tum.cit.aet.artemis.programming.repository.AuxiliaryRepositoryRepository;
+import de.tum.cit.aet.artemis.programming.repository.ProgrammingExerciseBuildConfigRepository;
+import de.tum.cit.aet.artemis.programming.repository.ProgrammingExerciseRepository;
+import de.tum.cit.aet.artemis.programming.service.GitService;
+import de.tum.cit.aet.artemis.programming.web.repository.FileSubmission;
+import de.tum.cit.aet.artemis.util.GitUtilService;
+import de.tum.cit.aet.artemis.util.LocalRepository;
+
+class AuxiliaryRepositoryResourceIntegrationTest extends AbstractSpringIntegrationJenkinsGitlabTest {
```
Comment (pzdr7) [inline src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:51]: I think #9314 broke the imports here. Can you have a look?

```
> Task :compileTestJava
/home/runner/work/Artemis/Artemis/src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:34: error: cannot find symbol
import de.tum.cit.aet.artemis.AbstractSpringIntegrationJenkinsGitlabTest;
                             ^
  symbol:   class AbstractSpringIntegrationJenkinsGitlabTest
  location: package de.tum.cit.aet.artemis
/home/runner/work/Artemis/Artemis/src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:48: error: package de.tum.cit.aet.artemis.util does not exist
import de.tum.cit.aet.artemis.util.GitUtilService;
                                  ^
/home/runner/work/Artemis/Artemis/src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:49: error: package de.tum.cit.aet.artemis.util does not exist
import de.tum.cit.aet.artemis.util.LocalRepository;
                                  ^
/home/runner/work/Artemis/Artemis/src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:51: error: cannot find symbol
class AuxiliaryRepositoryResourceIntegrationTest extends AbstractSpringIntegrationJenkinsGitlabTest {
                                                         ^
  symbol: class AbstractSpringIntegrationJenkinsGitlabTest
/home/runner/work/Artemis/Artemis/src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:76: error: cannot find symbol
    private final LocalRepository localAuxiliaryRepo = new LocalRepository(defaultBranch);
```

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Code looks good, but I think the tests aren't building (see below)

File: src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java (hunk 9 above)
Comment (SimonEntholzer) [reply src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:52]: Yep, the imports were indeed broken. I fixed them, now the tests work again :+1:

**Review by iyannsch — DISMISSED** [review]
Tested on TS3, works well even for empty aux. repos! Good job!

One thing: When the repo is empty, the file browser is also just empty - I'd love if we could show a `No files` or similar there ☺️
image

**Comment by SimonEntholzer** [comment]
> Tested on TS3, works well even for empty aux. repos! Good job!
> 
> One thing: When the repo is empty, the file browser is also just empty - I'd love if we could show a `No files` or similar there

I absolutely agree with you, but this is out of scope for this PR. In the future we will improve the repository view, with all its inner components.

File: src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+import de.tum.cit.aet.artemis.programming.util.ProgrammingExerciseFactory;
+import de.tum.cit.aet.artemis.programming.web.repository.FileSubmission;
+import de.tum.cit.aet.artemis.shared.base.AbstractSpringIntegrationJenkinsGitlabTest;
+
+class AuxiliaryRepositoryResourceIntegrationTest extends AbstractSpringIntegrationJenkinsGitlabTest {
+
+    private static final String TEST_PREFIX = "auxiliaryrepositoryresourceint";
+
+    @Autowired
+    private ProgrammingExerciseRepository programmingExerciseRepository;
```
Comment (pzdr7) [inline src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:57]: Could you use the `ProgrammingExerciseTestRepository` here?
```
    java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'classes should not use repositories with subclasses, because Test Repositories should be used over production repositories, if such a repository exist.' was violated (1 times):
  Test class de.tum.cit.aet.artemis.exercise.programming.AuxiliaryRepositoryResourceIntegrationTest uses repository de.tum.cit.aet.artemis.programming.repository.ProgrammingExerciseRepository which has subclasses: [JavaClass{name='de.tum.cit.aet.artemis.programming.test_repository.ProgrammingExerciseTestRepository'}]
```

Comment (SimonEntholzer) [reply src/test/java/de/tum/cit/aet/artemis/exercise/programming/AuxiliaryRepositoryResourceIntegrationTest.java:57]: good point, yes that works here :+1:

**Review by iyannsch — DISMISSED** [review]
Tested on TS3

**Review by pzdr7 — APPROVED** [review]
Code

File: src/main/webapp/app/detail-overview-list/detail-overview-list.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/webapp/app/detail-overview-list/detail-overview-list.component.html:7]: The point here is that $index is unique, whereas detail  is not in this case. detail can be "false" for multiple entires, which leads to errors complaining that it's not unique.

**Comment by krusche** [comment]
This PR is open for a long time. Let's please get it `ready to merge` asap!

**Review by pzdr7 — DISMISSED** [review]
Reapprove after merges

**Review by cremertim — APPROVED** [review]
Reapprove latest changes, tested on ts2

**Review by BBesrour — DISMISSED** [review]
tested on ts2, works as expected 👍

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

**Review by iyannsch — APPROVED** [review]

**Comment by SimonEntholzer** [comment]
I will increase the test coverage some more
